### DISCUSSION TITLE
Significant improvement on Identity Providers UI data loading - paged results

### DISCUSSION THREAD

**Discussion by laskasn** [keycloak/keycloak#8608] [Ideas]
### Motivation:

We have a Keycloak setup that needs to support multiple (in the order of thousands) (SAML) identity providers. When adding all of our  Identity providers (more than 4000), Keycloak has significant performance issues in the admin console especially when rendering the Identity Providers page. You can see in the [attached video](https://user-images.githubusercontent.com/11722418/137965926-e2b7cfa3-1d7b-460d-b5cc-ddf41db1eb6c.mp4) that it takes around half a minute to render the Identity Providers page and some seconds to render all others tabs. Some browsers become unresponsive in the Identity Providers page. Keycloak web services send in every page load  30,4ΜΒ of json serialized data ( 2 x 15.2) just to bring the realm information. Moreover, the admin Identity Providers page does not have a search box. Admins can only find the desired Identity Provider only with browser search. 

End-users also have performance issues during login as they are presented with a long list of all these IdPs in the login page.

### Enhancement description (our pull request):

In order to solve this problem, we made changes on two places: 

1. We removed from the **ModelToRepresentation.toRepresentation(RealmModel)** the inclusion of the IdentityProviders and the IdentityProviderMappers, so it transforms to a very light representation without these huge dangling sub-elements. We also added a **ModelToRepresentation.toBriefRepresentation(IdentityProviderModel)**, to be able generate a lighter version of the identity providers.
2. We added paging and search term parameters at the **IdentityProvidersResource.getIdentityProviders()** endpoint, to enable querying and fetching IdentityProviders in pages. We also made a small reform to the UI to list the IdentityProviders in pages, **exactly** as clients are listed.

Finally, we had to add 2 lines of code at ExportUtils.exportRealm(), to include explicitly the idps and the mappers which are not serialized in the representation.

With those 2 changes, the difference in the performance is outstanding, as it can be seen from the screenshots.

The UI, almost on each page, makes a request for the current realm and also for all the realms (realm[]) objects. So, if realm gets serialized with all its subelements packed in, it is very large. 

Realm configuration page stats (before and after the enhancement):

|                 | Before | After |
| ------------- | ------------- | ------------- |
| Total MBs needed to load | 33,3 mb  | 2,9 mb  |
| Time took to load data *  | 21,98 sec  | 1,39 sec  |

IdentityProviders listing page stats (before and after the enhancement):

|                 | Before | After |
| ------------- | ------------- | ------------- |
| Total MBs needed to load | 33,2 mb  | 2,9 mb  |
| Time took to load data *  | 30,9 sec  | 1,44 sec  |

\* on a typical 12mbit connection

In our keycloak cluster, we have only one realm loaded with IdPs. In the case that we load  another one realm with the same IdPs, the numbers on the "Before" column **will double**, while in the "After" column, will remain **almost the same**.    

You can find our pull request [here](https://github.com/keycloak/keycloak/pull/7960)

### Realms Before:

![realms_before-1](https://user-images.githubusercontent.com/11722418/137902922-3713a744-385e-40f6-a284-ca1e40d38aae.png)

### Realms after:

![realms_after-1](https://user-images.githubusercontent.com/11722418/137903084-19ab91e5-a4d2-45cf-aab9-0405dd325798.png)

### Identity providers before  **(notice the scrollbar)**:

![identityProviders_before-1](https://user-images.githubusercontent.com/11722418/137903209-efb0000a-f592-44ad-868f-180b66ec1e67.png)

### Identity Providers after:

![identityProviders_after-1](https://user-images.githubusercontent.com/11722418/137903323-6d4f6bbf-eaaa-4f9e-9f85-f3a62bef6dc1.png)

### Video of IdP loading prior applying the enhancement:

https://user-images.githubusercontent.com/11722418/137965926-e2b7cfa3-1d7b-460d-b5cc-ddf41db1eb6c.mp4

**Comment by stianst** [comment]
Adding support for pagination to identity providers is something we just should do. In contrast to the account console the admin console can and should just always show the pagination footer on the table.

**Reply by laskasn** [reply]
We have followed the exact same UI design with the clients' and the users' paging in the admin UI. In the Clients and Users in the admin UI listing, the "Next", "Previous" and "First page" buttons are not shown if no next page exists (results are less than 20). How do you belive that the "always show the pagination footer" should be implemented?  We could make the button disabled in the Identity provider listing, instead of hiding it whenever no next page exists if that's something we prefer - but again, wouldn't this confuse the Admin, since the other two listings {users and clients} follow the current design?
Let's move the discussion under the pull request, to further discuss the implementation details. Should we assign a reviewer for our [PR](https://github.com/keycloak/keycloak/pull/7960) ?

**Reply by stianst** [reply]
Never mind, I was mixing things up a little bit. We're doing the new admin console, which I believe always shows the footer, but if what you've done is consistent with users/clients in the old console that's great.

**Reply by laskasn** [reply]
Fine! Then, let's move on to the [pull request](https://github.com/keycloak/keycloak/pull/7960) then. Can we find someone @stianst to review it ? It would be great to have this upstream! You can find also the issue [here](https://issues.redhat.com/browse/KEYCLOAK-17860)

**Comment by laskasn** [comment]
@stianst Can we resume this PR? It's quite important for us - and probably for a significant part of the community.
It would be extremely helpful if we could have this in one of the following releases. We currently have this enhancement deployed on our keycloak clusters and we currently rebase our changes on top of the release, each time we want to upgrade to a newer release.
I think this enhancement will benefit a large part of the community who use more than a few IdPs on their keycloak(s).

**Comment by bonnm** [comment]
Hi,
@laskasn May I ask, how did you manage it to import more than 4000 SAML-IdPs into Keycloak? Are they part of a SAML-federation like eduGAIN? As far as I know, there is no support for automated SAML-federation xml-metadata processing in Keycloak. Even semi-automated/scripted import by admin API seems to be useless because of Keycloak's generated SAML-Brokering-URLs are individually generated per IdP, they contain the IdP-Alias as part of the path in the ACS-URL. So in my opinion, Keycloak cannot be registered directly as Service Provider in a SAML federation. I am very interested in how did you get this working...

**Comment by laskasn** [comment]
@bonnm Exactly! We have created a fork of keycloak (which we keep synchronized with the releases) in which we are able to load the eduGAIN federation. We have added an importer for SAML federations from a federation url (like [this one](https://md.aai.grnet.gr/feeds/edugain-idp-samlmd.xml) ) which synchronizes every X minutes. It loads all the IdPs and also applies mappers on each one of them. We use eduGAIN extensively in our production environments. I attach a few screenshots to showcase the additions from the vanilla keycloak. Our code resides in various branches in [our cloned repo](https://github.com/eosc-kc/keycloak) We were trying to convince the keycloak team that allowing keycloak to handle federations would be great, but they seem not to be intrigued that much. Hopefully in the future they'll change their minds. 
We have also modified the way it generates the SSPSSODescriptor for the SP side of the federation (keycloak is an SP for all eduGAIN IdPs) to be a single one for all IdPs ;-)  (i think this answers your question).
![edugain](https://user-images.githubusercontent.com/11722418/168675056-54811e4a-353f-4a42-87e1-dc7ba5014181.png)
![edugain_mappers](https://user-images.githubusercontent.com/11722418/168675077-cc565b8e-f260-4daf-a5a9-2621a9fd3cfb.png)
![edugain_details](https://user-images.githubusercontent.com/11722418/168675092-303b0f20-c0a8-4882-bc17-70d226db5f99.png)

**Reply by bonnm** [reply]
@laskasn Well that is incredible great work. It is totally unintelligible that this is not integrated in the main KC repo. This function opens KC to the whole research and education world. (I am from KIT (Karlsruhe, Germany) and I have colleagues working in the EOSC, but they are not involved with Keycloak). Your SAML federation work is not available as a standard [KC extension](https://www.keycloak.org/extensions.html) to use it in the normal version?
Let's hope this is integrated in KC (so we could get rid of these SAML-to-SAML extra bridges like satosa, simplesamlphp,... ), also the OIDC federation work you did! Keycloak is such a great tool but the not-availability of SAML fed is a killer argument against it in many cases I have seen...

**Reply by laskasn** [reply]
Well, the OIDC federation was quite an innovative work. We implemented the OpenID federation draft 13 (now it's at 18), which was integrated into keycloak (in our repo). But we have paused our work on that in order to finalize the SAML federation support.
Regarding your question now, there are technical issues that prevent it from becoming an extension (i think there's no point to list them here).
The idea is that we try to modify the core code as least as possible (it's inevitable to slightly modify the base code), in order to glue up our extended functionalities. 
However, we have different components which are provided as extensions, i.e. like [this one](https://github.com/rciam/keycloak-theme-vanilla). They follow the keycloak's SPI loading mechanism along with the wildfly's/quarkus' hotdeploy function.

**Reply by bonnm** [reply]
@laskasn Maybe you can start an new PR to integrate the SAML-federation in the main repo. There are many reservations against KC in the world of research and education, it's not well-accepted there. And the really only reason is the missing SAML federation (many other things are great compared to competitors)... Integrating such funtionality including OIDC federation would be much more useful than other things like q...

**Reply by laskasn** [reply]
I totally agree with you @bonnm 

We tried convincing keycloak developers in the past as i mentioned before, but at that time, we only had a simple prototype, while now we have a fully functional SAML federation-enabled keycloak running in our production machines and serving more than 4.500 IdPs.
I can prepare an extensive demonstration of what we have, but we would like you to express your interest about that and -if possible- get all other people also interested for this, join the conversation.  What do you think?

Well, let's not forget that we also need this discussion's PR ;)

**Reply by bonnm** [reply]
That's ok. If you do a new PR for this extension and the enhancements for account/admin console, I can participate in the corresponding discussions, expressing interest. But I can not participate in the PR itself, I'm not doing Java. (I even cannot run/test the new quarkus distribution, because we are on MS SQL and quarkus does not support newer versions of MS SQL, the included jbdc driver is years old...)
Next week I will have a talk about Keycloak in a widely spread german forum (education, research) with lots of IAM people and SAML experts in eduGAIN and the german sub federation (DFN-AAI). I yet prepared some slides about the federation problems of default Keycloak and your solution. I hope it's ok to show screenshots, I'm sure the audience's jaws drop when they see what you have...
